File: app/src/main/java/com/nononsenseapps/feeder/model/RssNotifications.java
/*
 * Copyright (c) 2014 Jonas Kalderstam.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nononsenseapps.feeder.model;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.support.v4.app.NotificationCompat;

import com.nononsenseapps.feeder.R;
import com.nononsenseapps.feeder.db.FeedItemSQL;
import com.nononsenseapps.feeder.db.FeedSQL;
import com.nononsenseapps.feeder.db.Util;
import com.nononsenseapps.feeder.ui.FeedActivity;
import com.nononsenseapps.feeder.ui.ReaderActivity;

import java.util.ArrayList;

/**
 * Handles notifications
 */
public class RssNotifications {

    private final static int mId = 73583;

    /**
     * Notify new items
     *
     * @param context
     */
    public static void notify(final Context context) {
        ArrayList<FeedItemSQL> feedItems = getItemsToNotify(context);

        NotificationManager nm =
                (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (feedItems.isEmpty()) {
            // Dismiss since it should be empty
            nm.cancel(mId);
            return;
        }

        NotificationCompat.Style notStyle;

        String contentTitle, contentText = "";
        // Many items
        if (feedItems.size() == 1) {
            notStyle = new NotificationCompat.BigTextStyle();
            contentTitle = "1 new RSS-story";
            ((NotificationCompat.BigTextStyle) notStyle).setBigContentTitle(contentTitle);
            contentText = feedItems.get(0).feedtitle + " \u2014 " + feedItems.get(0).plaintitle;
            ((NotificationCompat.BigTextStyle) notStyle).bigText(contentText);
        } else {
            notStyle = new NotificationCompat.InboxStyle();
            contentTitle = feedItems.size() + " new RSS-stories";
            ((NotificationCompat.InboxStyle) notStyle).setBigContentTitle(contentTitle);

            for (FeedItemSQL item : feedItems) {
                ((NotificationCompat.InboxStyle) notStyle).addLine(item.feedtitle + " \u2014 " + item.plaintitle);
                contentText += item.feedtitle + " \u2014 " + item.plaintitle + "\n";
            }
        }

        // Actions: markAsRead, enclosureplay, openlink
        // Priority: Low
        // Category: CATEGORY_SOCIAL
        // Style: INBOX

        Bitmap bm = BitmapFactory.decodeResource(context.getResources(), R.mipmap.ic_launcher);

        // TODO icon
        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(context)
                .setSmallIcon(R.drawable.ic_stat_rss)
                .setLargeIcon(bm)
                .setNumber(feedItems.size())
                .setAutoCancel(true)
                .setContentTitle(contentTitle)
                .setContentText(contentText)
                .setCategory(NotificationCompat.CATEGORY_SOCIAL)
                .setPriority(NotificationCompat.PRIORITY_LOW);

        if (feedItems.size() == 1) {
            if (feedItems.get(0).description != null && !feedItems.get(0).description.isEmpty()) {
                Intent intent = new Intent(context, ReaderActivity.class);
                ReaderActivity.setRssExtras(intent, feedItems.get(0));
                mBuilder.setContentIntent(PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT));
            } else {
                Intent intent = new Intent(context, FeedActivity.class);
                // TODO Set feed arguments
                mBuilder.setContentIntent(PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT));
            }

            if (feedItems.get(0).enclosurelink != null) {
                mBuilder.addAction(R.drawable.ic_action_av_play_circle_outline,
                        "Open file", PendingIntent.getActivity(context, 0, new Intent(Intent.ACTION_VIEW,
                                Uri.parse(feedItems.get(0).enclosurelink)), PendingIntent.FLAG_UPDATE_CURRENT));
            }
            mBuilder.addAction(R.drawable.ic_action_location_web_site,
                    "Open in browser", PendingIntent.getActivity(context, 0, new Intent(Intent.ACTION_VIEW,
                            Uri.parse(feedItems.get(0).link)), PendingIntent.FLAG_UPDATE_CURRENT));
        } else {
            Intent intent = new Intent(context, FeedActivity.class);
            mBuilder.setContentIntent(PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT));
        }
        // TODO?
        // mBuilder.addAction(R.drawable.ic_action_done_all, context.getString(R.string.mark_all_as_read), asreadpe);


        notStyle.setBuilder(mBuilder);

// mId allows you to update the notification later on.
        nm.notify(mId, notStyle.build());
    }

    public static ArrayList<FeedItemSQL> getItemsToNotify(final Context context) {
        ArrayList<FeedItemSQL> feedItems = new ArrayList<>();

        String feeds = getFeedsToNotify(context);

        if (feeds == null)
            return feedItems;

        Cursor c = context.getContentResolver()
                .query(FeedItemSQL.URI_FEED_ITEMS, FeedItemSQL.FIELDS,
                        FeedItemSQL.COL_FEED + " IN (" + feeds + ") AND " + FeedItemSQL.COL_NOTIFIED + " IS 0 AND " + FeedItemSQL.COL_UNREAD + " IS 1",
                        null, null);

        try {
            while (c.moveToNext()) {
                feedItems.add(new FeedItemSQL(c));
            }
        } finally {
            if (c != null) {
                c.close();
            }
        }

        return feedItems;
    }

    private static String getFeedsToNotify(Context context) {
        String result = null;

        Cursor c = context.getContentResolver()
                .query(FeedSQL.URI_FEEDS, Util.ToStringArray(FeedSQL.COL_ID),
                        FeedSQL.COL_NOTIFY + " IS 1",
                        null, null);

        try {
            while (c.moveToNext()) {
                if (result == null)
                    result = "";

                result += "" + Long.toString(c.getLong(0)) + ",";
            }
        } finally {
            if (c != null) {
                c.close();
            }
        }
        if (result != null)
            result = result.substring(0, result.length() - 1);
        return result;
    }
}

File: app/src/main/kotlin/com/nononsenseapps/feeder/model/RssNotifications.kt
package com.nononsenseapps.feeder.model

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.support.v4.app.NotificationCompat
import com.nononsenseapps.feeder.R
import com.nononsenseapps.feeder.db.*
import com.nononsenseapps.feeder.ui.FeedActivity
import com.nononsenseapps.feeder.ui.ReaderActivity
import com.nononsenseapps.feeder.util.getFeedItems
import com.nononsenseapps.feeder.util.getFeeds
import com.nononsenseapps.feeder.util.notificationManager


const val notificationId = 73583

fun notify(context: Context) {
    val feedItems = getItemsToNotify(context)

    val notification: Notification? = when (feedItems.size) {
        0 -> null
        1 -> singleNotification(context, feedItems.first())
        else -> manyNotification(context, feedItems)
    }

    val nm = context.notificationManager
    when (notification) {
        null -> nm.cancel(notificationId)
        else -> nm.notify(notificationId, notification)
    }
}

private fun singleNotification(context: Context, item: FeedItemSQL): Notification {
    val style = NotificationCompat.BigTextStyle()
    val title = item.plaintitle
    val text = item.plainsnippet

    style.bigText(text)
    style.setBigContentTitle(title)

    val intent = when (item.description.isBlank()) {
        true -> Intent(context, FeedActivity::class.java)
        false -> {
            val i = Intent(context, ReaderActivity::class.java)
            ReaderActivity.setRssExtras(i, item)
            i
        }
    }

    val builder = notificationBuilder(context)

    builder.setContentText(text)
            .setContentTitle(title)
            .setContentIntent(PendingIntent.getActivity(context, 0, intent,
                    PendingIntent.FLAG_UPDATE_CURRENT))

    if (item.enclosurelink != null) {
        builder.addAction(R.drawable.ic_action_av_play_circle_outline,
                context.getString(R.string.open_enclosed_media), PendingIntent.getActivity(context, 0, Intent(Intent.ACTION_VIEW,
                Uri.parse(item.enclosurelink)), PendingIntent.FLAG_UPDATE_CURRENT))
    }

    builder.addAction(R.drawable.ic_action_location_web_site,
            context.getString(R.string.open_link_in_browser),
            PendingIntent.getActivity(context, 0, Intent(Intent.ACTION_VIEW,
                    Uri.parse(item.link)), PendingIntent.FLAG_UPDATE_CURRENT))

    style.setBuilder(builder)
    return style.build()
}

private fun manyNotification(context: Context, feedItems: List<FeedItemSQL>): Notification {
    val style = NotificationCompat.InboxStyle()
    val title = context.getString(R.string.new_rss_stories)
    val text = feedItems.map { "${it.feedtitle} \u2014 ${it.plaintitle}" }.joinToString(separator = "\n")

    style.setBigContentTitle(title)
    feedItems.forEach {
        style.addLine("${it.feedtitle} \u2014 ${it.plaintitle}")
    }

    val intent = Intent(context, FeedActivity::class.java)

    val builder = notificationBuilder(context)

    builder.setContentText(text)
            .setContentTitle(title)
            .setContentIntent(PendingIntent.getActivity(context, 0, intent,
                    PendingIntent.FLAG_UPDATE_CURRENT))

    style.setBuilder(builder)
    return style.build()
}

private fun notificationBuilder(context: Context): NotificationCompat.Builder {
    val bm = BitmapFactory.decodeResource(context.resources, R.mipmap.ic_launcher)

    return NotificationCompat.Builder(context)
            .setSmallIcon(R.drawable.ic_stat_rss)
            .setLargeIcon(bm)
            .setAutoCancel(true)
            .setCategory(NotificationCompat.CATEGORY_SOCIAL)
            .setPriority(NotificationCompat.PRIORITY_LOW)
}

private fun getItemsToNotify(context: Context): List<FeedItemSQL> {
    val feeds = getFeedIdsToNotify(context)

    return when (feeds.isEmpty()) {
        true -> emptyList()
        false -> context.contentResolver.getFeedItems(
                where = "$COL_FEED IN (${feeds.joinToString(separator = ",")}) AND $COL_NOTIFIED IS 0 AND $COL_UNREAD IS 1")
    }
}

private fun getFeedIdsToNotify(context: Context): List<Long> =
        context.contentResolver.getFeeds(columns = listOf(COL_ID),
                where = "$COL_NOTIFY IS 1").map { it.id }